shared-module/audiofilters/Distortion.c

```diff
@@ -0,0 +1,409 @@
+            } else {
+                for (uint32_t i = 0; i < n; i++) {
+                    int32_t sample_word = 0;
+                    if (MP_LIKELY(self->bits_per_sample == 16)) {
+                        sample_word = sample_src[i];
+                    } else {
+                        if (self->samples_signed) {
+                            sample_word = sample_hsrc[i];
+                        } else {
+                            // Be careful here changing from an 8 bit unsigned to signed into a 32-bit signed
+                            sample_word = (int8_t)(((uint8_t)sample_hsrc[i]) ^ 0x80);
+                        }
+                    }
+
+                    // Apply pre-gain
+                    int32_t word = (int32_t)(sample_word * pre_gain);
+
+                    // Apply bit mask before converting to float
+                    if (self->mode == DISTORTION_MODE_LOFI) {
+                        word = word & word_mask;
+                    }
+
+                    if (self->mode != DISTORTION_MODE_LOFI || self->soft_clip) {
+                        // Convert sample to float
+                        mp_float_t wordf = word / MICROPY_FLOAT_CONST(32768.0);
+
+                        switch (self->mode) {
+                            case DISTORTION_MODE_CLIP: {
+                                wordf = MICROPY_FLOAT_C_FUN(pow)(MICROPY_FLOAT_C_FUN(fabs)(wordf), drive);
+                                if (word < 0) {
+                                    wordf *= MICROPY_FLOAT_CONST(-1.0);
+                                }
+                            } break;
+                            case DISTORTION_MODE_LOFI:
+                                break;
+                            case DISTORTION_MODE_OVERDRIVE: {
+                                wordf *= MICROPY_FLOAT_CONST(0.686306);
+                                mp_float_t z = MICROPY_FLOAT_CONST(1.0) + MICROPY_FLOAT_C_FUN(exp)(MICROPY_FLOAT_C_FUN(sqrt)(MICROPY_FLOAT_C_FUN(fabs)(wordf)) * MICROPY_FLOAT_CONST(-0.75));
+                                mp_float_t word_exp = MICROPY_FLOAT_C_FUN(exp)(wordf);
+                                wordf *= MICROPY_FLOAT_CONST(-1.0);
+                                wordf = (word_exp - MICROPY_FLOAT_C_FUN(exp)(wordf * z)) / (word_exp + MICROPY_FLOAT_C_FUN(exp)(wordf));
+                            } break;
+                            case DISTORTION_MODE_WAVESHAPE: {
+                                wordf = (MICROPY_FLOAT_CONST(1.0) + drive) * wordf / (MICROPY_FLOAT_CONST(1.0) + drive * MICROPY_FLOAT_C_FUN(fabs)(wordf));
+                            } break;
+                        }
+
+                        // Apply post-gain
+                        wordf = wordf * post_gain;
+
+                        // Soft clip
+                        if (self->soft_clip) {
+                            if (wordf > 0) {
+                                wordf = MICROPY_FLOAT_CONST(1.0) - MICROPY_FLOAT_C_FUN(exp)(-wordf);
```

Comment:
it's up to your testing that the perf is OK but I remain surprised that it's even feasible to call a transcendental math function for every sample. microcontrollers are a lot faster than they used to be.